Site Editor: Prepare route registration by refactoring the site editor router

youknowriad opened the pull request:
## What?

For a long time, people have been asking about an API to register pages and routes into the site editor. We're not ready yet to open such API because there are unknowns like: How to expand this API later to the full admin... That said, the current PR makes a huge progress towards that by adding an internal registration API and refactor the existing routes to use it.

## How?

A route object would be something like:

```
const myRoute = {
   name: "Name of the route",
   match: (params)=> boolean // function to see if the current url params match the current route
   areas: {
   },
   widths: {
   }
}
```

**Note**

This PR is not entirely ready, there are still some transitions to improve and we need to do some good testing and debugging. 

## Testing Instructions

Navigate everywhere in the site editor.

youknowriad committed:
Site Editor: Prepare route registration by refactoring the site editor router

youknowriad committed:
Better transition

youknowriad committed:
Memoize the active route

ramonjd commented on packages/edit-site/src/components/layout/router.js:
```diff
@@ -73,129 +61,26 @@ function useRedirectOldPaths() {
-				preview: <Editor />,
-				mobile: hasEditCanvasMode && <Editor />,
-			},
+			key: matchedRoute.name,
```
Is there a reason for the `name` vs `key` difference?

ramonjd reviewed (commented):
I am liking the direction of this centralized route registration! Great step to allow plugins etc to extend the site editor's routing capabilities.

I navigated back and forth between all sidebars routes, and, aside from the animations could not see any regressions.

There was one teensy difference between trunk and this PR: on the pages route, the data view panel is much wider.

| Trunk | This PR |
|--------|--------|
| <img width="400" alt="Screenshot 2024-10-14 at 10 27 07 am" src="https://github.com/user-attachments/assets/e1426250-c0b5-4c09-b8e2-f1ecf25892e1"> | <img width="400" alt="Screenshot 2024-10-14 at 10 27 11 am" src="https://github.com/user-attachments/assets/a167223f-a6b6-45d2-8239-a7fa302ecddb"> |

jsnajdr commented on packages/edit-site/src/components/layout/router.js:
```diff
@@ -73,129 +61,26 @@ function useRedirectOldPaths() {
 }
 
-export default function useLayoutAreas() {
+export default function useActiveRoute() {
```
nit: rename also the usage site which still calls this hook under the old `useLayoutAreas` name.

jsnajdr commented on packages/edit-site/src/components/site-editor-routes/navigation-edit.js:
```diff
@@ -0,0 +1,22 @@
+		return (
+			params.postType === NAVIGATION_POST_TYPE &&
+			! params.postId &&
+			params.canvas === 'edit'
```
Does the `canvas` mode really warrant a separate route? Aside from the `mobile` view it still renders the same sidebar and the same `<Editor />` component. The `canvas` mode is managed by the `canvasMode` state in the data store and synced by the `useSyncCanvasModeWithURL`.

In my view the canvas mode should be eventually managed only at one place; either:
a) in the route handler, which passes the mode down as `<Editor canvasMode={ params.canvas } />` prop;
b) inside the `<Editor />` component which reads the mode from the URL itself. This setup is very similar to how you would work with "minor" query params like `order=asc` or `search=foo`

youknowriad replied:
No disagreement, the main reason for the route is "mobile". I don't know how to handle it otherwise. For mobile, these routes render different things depending on the canvasMode. Not entirely sure how to handle it.

Aside: I would love to remove that duplicate state as well and only rely on the url but that is a separate large effort IMO.

mcsf replied:
I agree this is worth revisiting, although I should point out that I like the rigidity of the current proposal: other than the `match` predicate, everything else is static data, making the whole thing IMO very easy to keep track of. It's more difficult for surprising behaviour to creep in, especially once we start considering third-party use. (Note that this remark doesn't imply that addressing the `mobile` view in a smarter way would necessarily go against the principle.)

jsnajdr commented on packages/edit-site/src/store/reducer.js:
```diff
@@ -98,10 +98,20 @@ function editorCanvasContainerView( state = undefined, action ) {
+function routes( state = [], action ) {
+	switch ( action.type ) {
+		case 'REGISTER_ROUTE':
+			return [ ...state, action.route ];
```
We should probably also have a method to register multiple routes at once. Instead of having to go through `registry.batch`. Plugins will use it all the time.

Do we ever need to unregister a route?

youknowriad replied:
Unregistering seems like a good thing to have to potentially override a route or something.

jsnajdr reviewed (commented):
This looks nice. I wonder if we can make use of WordPress' `WP_Rewrite` to define pretty URLs for the wp-admin. Until now it's been used only to define permalink structure for the frontend, and also to do URL mapping for the REST API.

youknowriad commented:
> I wonder if we can make use of WordPress' WP_Rewrite to define pretty URLs for the wp-admin.

Would love that personally.

kevin940726 commented:
Sorry this is a high-level question that doesn't involve the code. Please ignore me if this has already been discussed elsewhere! Do you have any plans for integrating (or make compatible with) other third-party routing libraries in case we need more features? Even if we don't need it, I think it's nice to be aligned with popular libraries' API so that we know what we might run into in the future. Copying the API has some other advantages:

1. Familiar API for contributors with an established mental model.
2. Possible fork of documentation and/or linking to existing ones.
3. Battle-tested feature-complete API for unknown features in the future.
4. If we simply wrap an existing library then we can also receive regular updates for free.
5. Existing libraries usually work better with some other tools in the community (state management, testing, toolings, bundling, code splitting, preloading, lazy-loading...).

youknowriad commented:
> Do you have any plans for integrating (or make compatible with) other third-party routing libraries in case we need more features? Even if we don't need it, I think it's nice to be aligned with popular libraries' API so that we know what we might run into in the future.

Actually, this PR started locally by using a third-party router library (I tried both React Router and TanStack Router), I agree entirely that we should try using some of these libraries to avoid recreating the same feature over and over again. That said, here's what I found:

 - The router library used should be independent from the registration API we offer to third-party developers in WordPress, route registration as offered by these libraries is too low level and we want to provide something that is more specific with our areas well defined (sidebar, content, preview)
 - Using a third-party library is kind of independent from what this PR is trying to do. The router library would actually replace the existing `@wordpress/router` package instead and we will be offering this API as higher level registration for WP devs.

That said, as I said initially I did try to use one of the libraries to replace `@wordpress/router` as an experiment. I've discussed this at length with @jsnajdr and here's what we found:

These libraries are not easily adaptable to the use-case of multiple routable areas (sidebar, preview, content). The work by nesting routes within a single outlet (previous versions of react router support named outlets I think). So when using them it meant that we'd be "forced" to avoid using the nested capabilities which means losing nice transitions... between nested routes. Some potential solutions involved using "slots" and choosing one of our areas as the "outlet" area but it's too involved and not a clear "winner" solution. Another approach would be to use multiple routers for the different areas. I think these options (and maybe others) are still on the table but I wanted to focus on our immediate need first which is to address the higher-level registration problem.

youknowriad committed:
Cleanup

youknowriad committed:
Fix widths

youknowriad commented:
> I navigated back and forth between all sidebars routes, and, aside from the animations could not see any regressions.

For the animations, it's actually related to this comment https://github.com/WordPress/gutenberg/pull/66030#discussion_r1799657540 

Previously, the animation was triggered when we changed the route key/name or when we changed the canvasValue. With the current PR, these can happen both at the same time but since the "canvas" is a redux state value, it happens slightly after the route name change causing the animation to be triggered twice kind of which breaks it entirely. I made the choice to keep the animation only for canvas change for now but I think either removing the canvas state or using the same route for both canvasMode would solve the issue. But I think it's ok for now to not have animation when switching top level routes.

mcsf commented on packages/edit-site/src/components/site-editor-routes/index.js:
```diff
@@ -0,0 +1,60 @@
+	const { registerRoute } = unlock( useDispatch( siteEditorStore ) );
+	useEffect( () => {
+		registry.batch( () => {
+			routes.map( registerRoute );
```
```suggestion
			routes.forEach( registerRoute );
```

Here's my useless contribution ;)

mcsf reviewed (approved):
This looks nice, good enough to get it in and see how it feels.

youknowriad committed:
When does Miguel ever make a useful contribution?

Co-authored-by: Miguel Fonseca <150562+mcsf@users.noreply.github.com>

oandregal commented on packages/edit-site/src/components/site-editor-routes/home-view.js:
```diff
@@ -0,0 +1,16 @@
+	},
+	areas: {
+		sidebar: <SidebarNavigationScreenMain />,
+		preview: <Editor />,
```
I had forgotten why this didn't have a `mobile` area until I reviewed the `layout` component. Prepared https://github.com/WordPress/gutenberg/pull/66309 to document when each area is present. It also prevents the `edit` area from being rendered if the canvas mode is edit.
